crates/resolver-tests/src/lib.rs

```diff
@@ -411,22 +620,62 @@ impl SatResolver {
     }
 
     fn used_packages(&self) -> Option<String> {
-        self.0.borrow().solver.model().map(|lits| {
+        let s = self.0.borrow();
+        s.solver.model().map(|lits| {
             let lits: HashSet<_> = lits
                 .iter()
                 .filter(|l| l.is_positive())
                 .map(|l| l.var())
                 .collect();
-            let mut out = String::new();
-            out.push_str("used:\n");
-            for (p, v) in self.0.borrow().var_for_is_packages_used.iter() {
+
+            let mut used_packages = BTreeMap::<PackageId, BTreeSet<InternedString>>::new();
+            for (&p, v) in s.var_for_is_packages_used.iter() {
                 if lits.contains(v) {
-                    writeln!(&mut out, "    {}", p).unwrap();
+                    used_packages.entry(p).or_default();
+                }
+            }
+            for (&p, map) in &s.var_for_is_packages_features_used {
+                for (&f, v) in map {
+                    if lits.contains(v) {
+                        used_packages.entry(p).or_default().insert(f);
+                    }
                 }
             }
+
+            let mut out = String::from("used:\n");
+            for (package, feature_names) in used_packages {
+                writeln!(&mut out, "  {package}").unwrap();
+                for feature_name in feature_names {
+                    writeln!(&mut out, "    + {feature_name}").unwrap();
+                }
+            }
+
             out
         })
     }
+
+    #[cfg(test)]
+    fn reset_solver_lits(&self) {
```

Comment:
This is used for the test below, since a solution activating more packages than necessary is still valid.

Reply:
In practice, this seems to work, and be required. But it makes me a little nervous, the documentation for `.assume` says "This replaces the current set of assumed literals". So the API does not guarantee that  what's done in this method has any effect on subsequent calls to `sat_resolve`. I'm also not sure why it ends up being needed when `assumption.extend(old_root_vars.iter().map(|v| v.negative()))` is in `sat_resolve`.

What am I missing?

Reply:
I think when calling `solve` several times using [incremental solving](https://jix.github.io/varisat/manual/0.2.1/lib/incremental.html), the solver starts from the previous solution. Here I start with the `--all-features` case, which is also a solution of all the following cases (`default-features`, `--no-default-features`, `man` and `man,rgb`), so no step is needed for the solver and the variables are not updated. A call to `reset_solver_lits()` move away the starting point so it has more chances to give the preferred solution.

In any case this test is not very robust, since there is no guarantee that the solver will stop at the given set of dependencies and features. And the `resolve_and_validated` function only test the satisfiability property and not the solution found.

Reply:
> And the `resolve_and_validated` function only test the satisfiability property and not the solution found.

It's a little better than that. If cargo says there is no solution, then it checks that SAT thinks there's no solution. This is the best we can do in the no solution case (given the current resolver). If cargo says there is a solution, then we check that that solution is acceptable to SAT with `sat_is_valid_solution`.